File: 3191-minimum-operations-to-make-binary-array-elements-equal-to-one-i/3191-minimum-operations-to-make-binary-array-elements-equal-to-one-i.cpp
class Solution {
public:
    int minOperations(vector<int>& nums) {
        // Store indices of flip operations
        // Number of operations performed
        int ans = 0;
        queue<int> flipQueue;

        for (int i = 0; i < nums.size(); ++i){
            // Remove expired flips (older than 3 indices)
            while (!flipQueue.empty() && i > flipQueue.front() + 2)
                flipQueue.pop();

            // If the current element needs flipping
            if ((nums[i] + flipQueue.size()) % 2 == 0){
                // Cannot flip a full triplet
                if (i + 2 >= nums.size()) return -1;
                ++ans;
                flipQueue.push(i);
            }
        }

        return ans;
    }
};

File: 3191-minimum-operations-to-make-binary-array-elements-equal-to-one-i/3191-minimum-operations-to-make-binary-array-elements-equal-to-one-i.py
class Solution:
    def minOperations(self, nums: List[int]) -> int:
        # Store indices of flip operations
        # Number of operations performed
        queue, ans = deque(), 0

        for i in range(len(nums)):
            # Remove expired flips (older than 3 indices)
            while queue and i > queue[0] + 2:
                queue.popleft()
            
            # If the current element needs flipping
            if (nums[i] + len(queue)) % 2 == 0:
                # Cannot flip a full triplet
                if i + 2 >= len(nums):
                    return -1
                ans += 1
                queue.append(i)

        return ans